code/xcvm/cosmwasm/contracts/gateway/src/contract/sudo.rs

```diff
@@ -1,19 +1,42 @@
 use crate::error::ContractError;
-use cosmwasm_std::{entry_point, DepsMut, Env};
+use cosmwasm_std::{entry_point, wasm_execute, DepsMut, Env, Event, Response};
+use ibc_rs_scale::core::ics24_host::identifier::ChannelId;
 use xc_core::transport::ibc::{ics20::hook::IBCLifecycleComplete, SudoMsg};
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn sudo(_deps: DepsMut, _env: Env, msg: SudoMsg) -> crate::error::Result {
+pub fn sudo(deps: DepsMut, env: Env, msg: SudoMsg) -> crate::error::Result {
+	deps.api
+		.debug(&format!("cvm::gateway::sudo {}", serde_json_wasm::to_string(&msg)?));
 	match msg {
 		SudoMsg::IBCLifecycleComplete(IBCLifecycleComplete::IBCAck {
-			channel: _,
-			sequence: _,
-			ack: _,
-			success: _,
-		}) => Err(ContractError::NotImplemented),
-		SudoMsg::IBCLifecycleComplete(IBCLifecycleComplete::IBCTimeout {
-			channel: _,
-			sequence: _,
-		}) => Err(ContractError::NotImplemented),
+			channel,
+			sequence,
+			ack,
+			success,
+		}) =>
+			if !success {
+				handle_transport_failure(deps, env, channel, sequence, ack)
+			} else {
+				Ok(Response::new())
+			},
+		SudoMsg::IBCLifecycleComplete(IBCLifecycleComplete::IBCTimeout { channel, sequence }) =>
+			handle_transport_failure(deps, env, channel, sequence, "timeout".to_string()),
 	}
 }
+
+/// return funds to interpreter and sets final error
+fn handle_transport_failure(
```

Comment:
why not pass the `msg` as argument instead?

Reply:
because message is both success and failure and has different interface and SUDO can be other then IBC.

so for handling just IBC failrues there is only specific data taken from context.